substrate/frame/support/src/traits/storage.rs

```diff
@@ -206,16 +206,20 @@ where
 /// treated as one*. Don't type to duplicate it, and remember to drop it when you're done with
 /// it.
 #[must_use]
-pub trait Consideration<AccountId>: Member + FullCodec + TypeInfo + MaxEncodedLen {
+pub trait Consideration<AccountId, Footprint>:
+	Member + FullCodec + TypeInfo + MaxEncodedLen
+{
 	/// Create a ticket for the `new` footprint attributable to `who`. This ticket *must* ultimately
-	/// be consumed through `update` or `drop` once the footprint changes or is removed.
-	fn new(who: &AccountId, new: Footprint) -> Result<Self, DispatchError>;
+	/// be consumed through `update` or `drop` once the footprint changes or is removed. `None`
+	/// implies no cost for a given footprint.
+	fn new(who: &AccountId, new: Footprint) -> Result<Option<Self>, DispatchError>;
```

Comment:
This seems to be bleeding the implementation details outside. Basically, the caller should know that the consideration being None means that the storage cost was 0. 

Instead, shouldn't the consideration expose an API to check whether it has some underlying storage or not?

Reply:
> This seems to be bleeding the implementation details outside. Basically, the caller should know that the consideration being None means that the storage cost was 0.

Does not the client already know that the ticket is about the storage cost? If you check the doc for this trait, it does expect a ticket to carry an information about a cost. Or I misunderstand what you mean. 

> Instead, shouldn't the consideration expose an API to check whether it has some underlying storage or not?

Not sure I understand this part. Does not the client of the trait the one how knows about a storage? What kind of API you mean, can you please give an example.

My example is:

``` rust

type Proposals = StorageMap<...>;
type ProposalDeposit = StorageMap<.., ProposalIndex, T::Consideration::Ticket...>;
...
pub fn submit_proposal(...) {
  ..
  // for example if proposal_count < 10, no cost, if greater some cost.
  let ticket = T::Consideration::new(proposal_count);
  // if ticket cannot be none, we have to always save a ticket with zero deposit.
  ProposalDeposit::insert(proposal_index, ticket);
}
```

Not bid deal if the ticket itself can mean a zero cost (with zero balance for example), but a none option looks better to me.

Reply:
> Does not the client already know that the ticket is about the storage cost? If you check the doc for this trait, it does expect a ticket to carry an information about a cost. Or I misunderstand what you mean.

So, the caller already has access to the footprint and hence, can already know that it doesn't require a ticket. However, we still try to obtain a consideration for it. In this case, it returns None which is then used to signify that the ticket doesn't need to be saved. May be I misunderstood something, but this sounds like a round about way to achieve something that could be done at the first step itself? I mean what's the point of trying to obtain a Consideration here?

Reply:
@gupnik the client (e.g. a pallet) of the `Consideration` does not know if there is a cost for a certain `footprint`. The `Consideration` provider (e.g. a pallet's client) defines the rules.

Reply:
> @gupnik the client (e.g. a pallet) of the `Consideration` does not know if there is a cost for a certain `footprint`. The `Consideration` provider (e.g. a pallet's client) defines the rules.

Ahh, makes sense! Thank you.